### DISCUSSION TITLE
Webhook Failed debug

### DISCUSSION THREAD

**Discussion by dineshwar21** [emqx/emqx#13858] [Q&A]
I've integrated a webhook with an HTTP POST endpoint, but some of the webhook requests are failing intermittently. When I inspect the EMQX API metrics for the webhook , the failure reason is listed as 'unknown,' with no specific error details provided. Is there a way to investigate or debug further to identify the root cause of these failures?
![image](https://github.com/user-attachments/assets/5719efb9-fa28-4691-a018-d74ac79c7116)

**Comment by keynslug** [comment]
Hi @dineshwar21,

I'd expect you could see error-level events in the EMQX logs in this case, please share if you find any. If not, I guess this [feature](https://docs.emqx.com/en/emqx/latest/observability/tracer.html#trace-by-rule-id) can potentially help but it might be hard to hit it given how rare these event are.

**Comment by dineshwar21** [comment]
EMQX webhook responses are not in order, causing unordered data. With more than 15 responses per second, is there a better way to handle this?

**Reply by dineshwar21** [reply]
There is a delay in the response timestamp and response log time it's in UTC

2024-11-07 10:02:56:{"publish_received_at":1731016961059,"pub_props":{"User-Property":{}},"qos":1,"topic":"test/topic","clientid":"199999","client_attrs":{},"peerhost":"","payload":"PAYLOAD","username":"test","event":"message.publish","metadata":{"rule_id":"API_RESPONSE_WH_D"},"peername":"","timestamp":1731016961059,"node":"emqx@127.0.0.1","id":"00062659CB106AD4F4450066D1C501EF","flags":{"retain":false,"dup":false}}

publish_received_at = November 7, 2024 10:02:41.059 PM

**Reply by dineshwar21** [reply]
I’m using a webhook for a functionality, but sometimes the webhook data doesn't respond in a timely manner and some get failed. I need to understand how it works internally and if there are any limitations to it.

Will there be a delay in response delivery?
Webhook metrics 
![image](https://github.com/user-attachments/assets/b91625c7-379c-4583-943a-d50c4ff3892f)

**Reply by zmstone** [reply]
Hi @dineshwar21 
if you open development tools in the browser, you'll see sub groups of the "Dropped" counters.
Most common reason for a message to be dropped is the request TTL, if the request stayed in the send buffer for longer than request TTL, it's dropped.
Another common reason is when the send buffer is full.

`X MQTT client processes ---> N buffers ---> M HTTP clients ---> Webhook Server`

The buffer is where the counters are bumped. HTTP clients may reply errors to buffers,
then the buffers map the errors into counters.

**Reply by dineshwar21** [reply]
@zmstone 

Thank you for your time. 
1. Will it take 15 seconds for a webhook response to be delivered to the server? 
2. This is the buffer value set in the webhook
![image](https://github.com/user-attachments/assets/902bc053-e830-4c98-8050-c0223c8b0368)
3. Request TTL also enabled.
![image](https://github.com/user-attachments/assets/9d8c8726-60ec-4770-b719-e807e41a09d2)

Note, I have two webhooks: one (Webhook1) receives an average of 5 to 10 messages per second, and the other (Webhook2) which got 15-second response delay. Has 0.7 messages per second. Will Webhook1 cause a delay in Webhook2?

Below is the config which i have. Do i need to change anything. So i can get message instantly
![image](https://github.com/user-attachments/assets/7e2e58ef-11fc-4e69-8df0-13020a81ca52)

**Reply by zmstone** [reply]
If webhook1 and webhook2 share the same connector, they will affect each other.

caveat: EMQX webhook client keeps the underlying TCP connection open to stream requests,
due to the nature of HTTP protocol, what is sent can never be recalled, for example, if a request is sent, there is no way for the client to revoke the request from the server, there is also no "deadline" concept in HTTP protocol.
This means: if the server may silently drop an HTTP request (no 400+ or 500+ reply, no TCP socket close etc), then the client will experience timeout.
Before EMQX 5.8.1, such timeout is `infinity`, after 5.8.1, webhook client will reconnect after `TTL+10s`, and the expired requests will be dropped before retry.

**Reply by dineshwar21** [reply]
Will there be a delay in message delivery. One message took 15 seconds to delivery after it publish to MQTT broker.

Same connector means same node?

**Reply by zmstone** [reply]
If the action is sync mode, yes, it delays the message publish.
If the action is async mode, then no

**Reply by dineshwar21** [reply]
I have set Query Mode to async only.

**Reply by dineshwar21** [reply]
Do any configuration changes will fix the issue.

**Reply by zmstone** [reply]
Hi @dineshwar21 
In this thread, I have answered some questions about latency, and how EMQX webhook integration works internally.

I'd like to get more clarification with regards to what exact issue you are referring to.

- In the original post, it was about `unknown` failures. (for this we need to find the clues in the logs, but have not got any logs from you)
- In this thread, I advised to inspect the detailed "drop" counter,  but have not got a response from you.
> if you open development tools in the browser, you'll see sub groups of the "Dropped" counters.
- In this thread, you asked about ordering. I tried to read through the thread again, but I am afraid I still need more clarification

**Reply by dineshwar21** [reply]
@zmstone, Thanks for your time.

I have two webhooks, and the order in which they are triggered is important because the hooks are dependent on each other. One webhook receives 8 to 10 responses per second, while the other rarely receives a response. Since the second webhook's responses are of higher priority, it is processed separately. However, the problem is that the response from the second webhook is delayed by 15 seconds. Although the MQTT broker receives the response, it is only delivered to the webhook after a 15-second delay.

**Reply by zmstone** [reply]
I see what you mean now.
If a message matches multiple rules, the execution order of the rules is not defined.

An alternative is to change from webhook to "http bridge".

- create two http connectors, one for each of your current webhook server/endpoint
- create one rule which can produce query results for both http integrations
- add two http actions for the same rule
  - make sure the first action has higher priority, and choose "sync" for "Query Mode"
  - the second action can be the slow one, for which you can use "async" mode if you do not wish to cause back-pressure toward the MQTT client.

This should ensure that the two actions are not triggered concurrently for messages from one client. 
This however requires you to always trigger both actions at the same time because there is no condition/filter in the actions (only in the rules)
